subprojects/core/src/main/java/org/gradle/api/capabilities/CapabilitiesExtension.java

```diff
@@ -0,0 +1,168 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.gradle.api.capabilities;
+
+import com.google.common.base.Objects;
+import com.google.common.collect.ArrayListMultimap;
+import com.google.common.collect.Multimap;
+import org.gradle.api.Incubating;
+import org.gradle.api.InvalidUserDataException;
+import org.gradle.api.artifacts.Configuration;
+import org.gradle.api.artifacts.ConfigurationContainer;
+import org.gradle.internal.metaobject.DynamicInvokeResult;
+import org.gradle.internal.metaobject.MethodAccess;
+import org.gradle.internal.metaobject.MethodMixIn;
+
+import java.util.Collection;
+
+/**
+ * The project extension responsible for managing the capabilities of a component.
+ *
+ * @since 4.7
+ */
+@Incubating
+public class CapabilitiesExtension implements MethodMixIn {
+    private final ConfigurationContainer configurations;
+    private final MethodAccess dynamicMethods;
+
+    private final Multimap<Configuration, CapabilityDescriptor> descriptors = ArrayListMultimap.create();
+
+    public CapabilitiesExtension(ConfigurationContainer configurations) {
+        this.configurations = configurations;
+        this.dynamicMethods = new CapabilitiesExtensionMethodAccess();
+    }
+
+    public CapabilityDescriptor add(Configuration configuration, String notation) {
+        String[] parts = validateNotation(notation);
+        return add(configuration, parts[0], parts[1], parts[2]);
+    }
+
+    public CapabilityDescriptor add(String configurationName, String group, String name, String version) {
+        return add(configurations.getByName(configurationName), group, name, version);
+    }
+
+    public CapabilityDescriptor add(Configuration configuration, String group, String name, String version) {
```

Comment:
I think we should aim for a consistent API with the `dependencies` block. I don't see why we can't leverage the existing `NotationParser` infrastructure to do this.